src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,180 @@
+               std::numeric_limits<size_t>::max() : maxIterations;
+
+  // initialize the gradients.
+  size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+  arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+  arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+  for (size_t f = 0, b=0; f < numFunctions;
+    /* incrementing done manually */)
+  {
+    // Find the effective batch size (the last batch may be smaller).
+    effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+    function.Gradient(iterate, f, tableOfGradients.slice(b),
+                      effectiveBatchSize);
+
+    f += effectiveBatchSize;
+    avgGradient += tableOfGradients.slice(b);
+    b++;
+  }
+  avgGradient /= (double) numBatches; // average gradient
+
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "SAGA: converged to " << overallObjective
+                << "; terminating  with failure.  Try a smaller step size?"
+                << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SAGA: minimized within tolerance " << tolerance
+                << "; terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    lastObjective = overallObjective;
+
+    // Store current parameter for the calculation of the variance reduced
+    // gradient.
+    iterate0 = iterate;
+
+    for (size_t f = 0, b=0, currentFunction = 0; f < numFunctions;
+      /* incrementing done manually */)
+    {
+      b = math::RandInt(0, numBatches); // Random batch selected
+      // Find the effective batch size (the last batch may be smaller).
+      currentFunction = b*batchSize;
+      effectiveBatchSize = std::min(batchSize, numFunctions - currentFunction);
+
+      // Is this iteration the start of a sequence?
+      if ((currentFunction % numFunctions) == 0)
+      {
+        currentFunction = 0;
+
+        // Determine order of visitation.
+        if (shuffle)
+          function.Shuffle();
+      }
+
+      // Calculate the gradient of a random function.
+      function.Gradient(iterate, currentFunction, gradient,
+                        effectiveBatchSize);
+      gradient0 = tableOfGradients.slice(b);
+
+      // Use the update policy to take a step.
+      updatePolicy.Update(iterate, avgGradient, gradient, gradient0,
+                          stepSize);
+      // update average
+      avgGradient += (gradient-tableOfGradients.slice(b))/numBatches;
+
+      // update table of gradients
```

Comment:
It's helpful to always have comments that are complete sentences with proper grammar and punctuation; see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#comments for more information.

Reply:
Sorry, I'll keep that in mind for future!